### PR TITLE
Fix trainer subscription

### LINKED ISSUE #85: Trainer pubsub subscription retries thousands of times before getting a response.
There's an error that occurs while running the trainer service, where some file isn't found where it's expected.
- Pub sub subscriptions are 'outstanding' until they receive an ack response (see [here](https://cloud.google.com/pubsub/docs/subscriber))
- If no ack response is received within a set time (defaults to 10 seconds), the request is retried until it succeeds.  
- Currently the trainer service waits for the entire training to complete before returning a response.
- This means that although the cloud run service was running, the pubsub subscription had no way to verify that it's request had been received, and was sending tonnes and tonnes of requests to the cloudrun server. 
- Obviously this is a massive issue if we're paying for big compute resources to run the docker images, and they continue to run forever because they get spammed with requests.

### PR DESCRIPTION
**Pull request by harrykeightley**
Fixes #85 

- Makes trainer service return an ack response as soon as it assembles a valid `ModelMetadata` instance.
- Update the google provider versions in terraform to enable `exactly_once_delivery` on the topic subscription.
- Increase the `ack_deadline_seconds` to 30. 
- Add tests for hitting `/` with various request formats. 
- Minor refactor of logging and responses in trainer service.

### CODE REVIEW

**Review by nicklambourne — APPROVED** [review]
LGTM w/ suggestions

File: services/trainer/trainer/main.py [hunk 1]
```diff
@@ -3,64 +3,86 @@
...
+        data = json.loads(data)
         metadata = ModelMetadata.from_dict(data)
     except Exception:
-        msg = f"Invalid data format: {data}"
-        logger.error(f"Error: {msg}")
-        return f"Bad Request: {msg}", 400
+        msg = f"{BAD_MODEL_METADATA_FORMAT}: {data}"
+        return bad_request(msg)
+
+    response = Response(status=204)
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/main.py]: Fixed in latest revision.

File: services/trainer/trainer/main.py [hunk 2]
```diff
@@ -3,64 +3,86 @@
...
+    # Queue the training to happen after returning an ACK response for the
+    # pubsub subscription, so it doesn't keep retrying.
+    response.call_on_close(lambda: process_training_request(metadata))
+
+    logger.info("Responding with ACK")
+    return response
+
+
+def bad_request(msg: str) -> Response:
+    """Logs the given error message and builds a 400 response to return.
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/main.py:62]: Hey I'm a little confused about this- if the docstring says it builds a 400 response, what other status codes would be valid? Are you saying that "400 response" implies any 40X status code?

[the remark this answers is not in the document]
Comment (nicklambourne) [reply services/trainer/trainer/main.py:62]: Did this, when I originally reviewed it, allow for the user to pass an arbitrary status code, but default to 400? I'd look it up myself, but rebase + force push means I can't (which is why I normally discourage people from using it).

[the remark this answers is not in the document]
Comment (nicklambourne) [reply services/trainer/trainer/main.py:62]: As it is now it's fine and there's no contradiction.

[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/main.py:62]: That's a good point about rebasing that I didn't consider haha - I've been thinking a lot this past week about the optimal way to keep a clean git history and I think I've come round to the sqash merge (which is what you originally set up) and keep all the commits visible on the branch in the mean time.

[the remark this answers is not in the document]
Comment (harrykeightley) [reply services/trainer/trainer/main.py:62]: As for the function sig, I only changed the wording of `msg` to `error_msg` haha

[the remark this answers is not in the document]
Comment (nicklambourne) [reply services/trainer/trainer/main.py:62]: I must be hallucinating, don't mind me 😅

**Pull request merged**